alf/utils/common.py

```diff
@@ -440,23 +453,28 @@ def _traverse_path(obs, path):
         if not path:
             return _transform_image(obs)
         step = path[0]
-        if isinstance(obs, tuple) and hasattr(obs, '_fields'):
-            new_val = _traverse_path(getattr(obs, step), path[1:])
-            return obs._replace(**{step: new_val})
+        if isinstance(obs, tuple):
+            if hasattr(obs, '_fields'):
+                new_val = _traverse_path(getattr(obs, step), path[1:])
+                return obs._replace(**{step: new_val})
+            else:
+                step = int(step)
+                transformed = _traverse_path(obs[step], path[1:])
+                return obs[:step] + (transformed, ) + obs[step + 1:]
```

Comment:
This will fail if obs is a list

Reply:
Honestly I don't think supporting unnamed tuple is a good choice. Basically the user has to provide a sequence of integer indices which might be error-prone.

Reply:
there  is a problem in tf_agent EncodingNetwork that can not take  namedtuple or OrderDict observation as input , so add support for unnamed tuple for now

Reply:
OK, remember to update the function comment

Reply:
> there is a problem in tf_agent EncodingNetwork that can not take namedtuple or OrderDict observation as input , so add support for unnamed tuple for now

EncodingNetwork does support dict observation:
https://github.com/HorizonRobotics/alf/pull/205/files#diff-4f8caad82255bf76df66e98c7f055370

Reply:
Yes, it supports, i made a mistake before (with a wrong preprocessing_layers) . Revert to the previous implementation in the new commit